src/t/d20.clj

```diff
@@ -0,0 +1,260 @@
+(ns t.d20
+  (:gen-class)
+  (:require
+    [clojure.string :as cljstr]
+    [clojure.set :as cljset]
+    [clojure.core.match :refer [match]]))
+
+(def sample
+"###############
+#...#...#.....#
+#.#.#.#.#.###.#
+#S#...#.#.#...#
+#######.#.#.###
+#######.#.#...#
+#######.#.###.#
+###..E#...#...#
+###.#######.###
+#...###...#...#
+#.#####.#.###.#
+#.#...#.#.#...#
+#.#.#.#.#.#.###
+#...#...#...###
+###############")
+
+(defn parse-input
+  [input]
+  (let [lines (cljstr/split-lines input)
+        ymax (count lines)
+        xmax (count (first lines))
+        all-map (->> lines
+                     (map-indexed
+                       (fn [y l]
+                         (apply merge-with into
+                                (map-indexed
+                                  (fn [x c]
+                                    (case c
+                                      \S {:start [[x y]]}
+                                      \# {:walls [[x y]]}
+                                      \E {:end   [[x y]]}
+                                      {}))
+                                  l))))
+                     (apply merge-with into)) ]
+    {:walls (set (:walls all-map))
+     :S (first (:start all-map))
+     :E (first (:end all-map))
+     :xmax xmax
+     :ymax ymax}
+  ))
+
+
+(defn inside?
+  [x y xmax ymax]
+  (and (< -1 x xmax) (< -1 y ymax)))
+
+(defn def-find-shortest-path
+  [walls E xmax ymax]
+  (fn [S]
+    (loop [p [S]
+           N 0]
+        (let [np (mapcat
+                   (fn [[x y]] [[(inc x) y] [(dec x) y] [x (inc y)] [x (dec y)]])
+                   p)
+              np (->> np
+                      (filter (fn [[x y]] (inside? x y xmax ymax)))
```

Comment:
If you redefine `inside?` to be something like:

```clojure
(defn inside?
  [width height]
  (let [xmax (dec width)
        ymax (dec height)]
  (fn [[x y]]
    (and (<= 0 x xmax) (<= 0 y ymax)))))
```

you can write:

```suggestion
                      (filter (inside? xmax ymax))
```

Reply:
yes, I am slowly starting to use `fn` but it is in progress